stage1/init/kvm.go

```diff
@@ -113,16 +115,39 @@ func mountSharedVolumes(root string, p *stage1commontypes.Pod, ra *schema.Runtim
 }
 
 func doBindMount(source, destination string, readOnly bool, recursive *bool) error {
-	var flags uintptr = syscall.MS_BIND | syscall.MS_REC
+	var flags uintptr = syscall.MS_BIND
+
 	if readOnly {
 		flags |= syscall.MS_RDONLY
 	}
 
 	// Enable recursive by default and remove it if explicitly requested
-	if recursive != nil && *recursive == false {
-		flags ^= syscall.MS_REC
+	recursiveBool := recursive == nil || *recursive == true
+	if recursiveBool {
+		flags |= syscall.MS_REC
+	}
+
+	if err := syscall.Mount(source, destination, "bind", flags, ""); err != nil {
+		return err
 	}
-	return syscall.Mount(source, destination, "bind", flags, "")
+
+	if readOnly && recursiveBool {
+		// Sub-mounts are still read-write, so find them and remount them read-only
+
+		mnts, err := stage0.GetMountsForPrefix(source + "/")
+		if err != nil {
+			return errwrap.Wrap(fmt.Errorf("error getting mounts under %q from mountinfo", source), err)
+		}
+
+		for _, mnt := range mnts {
+			innerAbsPath := destination + strings.Replace(mnt.MountPoint, source, "", -1)
+			if err := syscall.Mount("", innerAbsPath, "none", syscall.MS_REMOUNT|syscall.MS_RDONLY, ""); err != nil {
+				return err
```

Comment:
I tested with the additional flag `MS_BIND`:
```
-                       if err := syscall.Mount("", innerAbsPath, "none", syscall.MS_REMOUNT|syscall.MS_RDONLY, ""); err != nil {
+                       if err := syscall.Mount("", innerAbsPath, "none", syscall.MS_REMOUNT|syscall.MS_BIND|syscall.MS_RDONLY, ""); err != nil {
```

Then, I tried again:
```
sudo rkt run --debug  --stage1-path=/usr/bin/stage1-kvm.aci --insecure-options=image,ondisk --volume=foo,readOnly=true,recursive=true,kind=host,source=/tmp/data docker://busybox --mount=volume=foo,target=/data --exec /bin/sh -- -c 'touch /data/KVM ; touch /data/RO/KVM ; touch /data/RW/KVM'
```

This time, rkt can start correctly. But the mount is not read-only as expected because KVM managed to create the file:
```
$ find /tmp/data -name KVM
/tmp/data/KVM
```

This should be added in the tests.